pkg/flow/internal/controller/node_service.go

```diff
@@ -43,14 +52,73 @@ func (sn *ServiceNode) NodeID() string { return sn.def.Name }
 // Block implements BlockNode. It returns nil, since ServiceNodes don't have
 // associated configs.
 func (sn *ServiceNode) Block() *ast.BlockStmt {
-	// TODO(rfratto): support configs for services.
-	return nil
+	sn.mut.RLock()
+	defer sn.mut.RUnlock()
+	return sn.block
+}
+
+// UpdateBlock updates the River block used to construct arguments for the
+// service. The new block isn't used until the next time Evaluate is called.
+//
+// UpdateBlock will panic if the block does not match the ID of the
+// ServiceNode.
+//
+// Call UpdateBlock with a nil block to remove the block associated with the
+// ServiceNode.
+func (sn *ServiceNode) UpdateBlock(b *ast.BlockStmt) {
+	if b != nil && !BlockComponentID(b).Equals([]string{sn.NodeID()}) {
+		panic("Updateblock called with a River block with a different block ID")
+	}
+
+	sn.mut.Lock()
+	defer sn.mut.Unlock()
+
+	sn.block = b
+
+	if b != nil {
+		sn.eval = vm.New(b.Body)
+	} else {
+		sn.eval = vm.New(ast.Body{})
+	}
 }
 
 // Evaluate implements BlockNode. It is a no-op since ServiceNodes don't have
 // associated configs to evaluate.
 func (sn *ServiceNode) Evaluate(scope *vm.Scope) error {
-	// TODO(rfratto): support configs for services.
+	sn.mut.Lock()
```

Comment:
Just making sure; since services don't have the concept of health, we're depending on loading them correctly the first time around and then rejecting any invalid config, right?

Reply:
> Just making sure; since services don't have the concept of health, we're depending on loading them correctly the first time around and then rejecting any invalid config, right?

In the current implementation, they're still technically allowed to run (at least from the perspective of the loader) even if they fail to evaluate. We still need more work to tighten this down, especially with #4411 being a candidate for merging.